apps/confidential/confidential_test.go

```diff
@@ -204,55 +204,79 @@ func TestAcquireTokenByAuthCode(t *testing.T) {
-			UID:  "123-456",
-			UTID: "fake",
-		},
-	}, cred)
-	if err != nil {
-		t.Fatal(err)
-	}
-	_, err = client.AcquireTokenSilent(context.Background(), tokenScope)
-	// first attempt should fail
-	if err == nil {
-		t.Fatal("unexpected nil error from AcquireTokenSilent")
-	}
-	tk, err := client.AcquireTokenByAuthCode(context.Background(), "fake_auth_code", "fake_redirect_uri", tokenScope)
-	if err != nil {
-		t.Fatal(err)
-	}
-	if tk.AccessToken != token {
-		t.Fatalf("unexpected access token %s", tk.AccessToken)
-	}
-	account := client.Account(tk.Account.HomeAccountID)
-	// second attempt should return the cached token
-	tk, err = client.AcquireTokenSilent(context.Background(), tokenScope, WithSilentAccount(account))
-	if err != nil {
-		t.Fatal(err)
-	}
-	if tk.AccessToken != token {
-		t.Fatalf("unexpected access token %s", tk.AccessToken)
+	for _, params := range []struct {
+		upn, preferredUsername, utid string
+	}{
+		{"", "fakeuser@fakeplace.fake", "fake"},
+		{"fakeuser@fakeplace.fake", "", ""},
+	} {
+		t.Run("", func(t *testing.T) {
+			tr := accesstokens.TokenResponse{
+				AccessToken:   token,
+				RefreshToken:  refresh,
+				ExpiresOn:     internalTime.DurationTime{T: time.Now().Add(1 * time.Hour)},
+				ExtExpiresOn:  internalTime.DurationTime{T: time.Now().Add(1 * time.Hour)},
+				GrantedScopes: accesstokens.Scopes{Slice: tokenScope},
+				IDToken: accesstokens.IDToken{
+					PreferredUsername: params.preferredUsername,
+					UPN:               params.upn,
+					Name:              "fake person",
+					Oid:               "123-456",
+					TenantID:          "fake",
+					Subject:           "nothing",
+					Issuer:            "https://fake_authority/fake",
+					Audience:          "abc-123",
+					ExpirationTime:    time.Now().Add(time.Hour).Unix(),
+					IssuedAt:          time.Now().Add(-5 * time.Minute).Unix(),
+					NotBefore:         time.Now().Add(-5 * time.Minute).Unix(),
+					// NOTE: this is an invalid JWT however this doesn't cause a failure.
+					// it simply falls back to calling Token.Refresh() which will obviously succeed.
+					RawToken: "fake.raw.token",
+				},
+				ClientInfo: accesstokens.ClientInfo{
+					UID:  "123-456",
+					UTID: params.utid,
+				},
+			}
+
+			client, err := fakeClient(tr, cred)
+			if err != nil {
+				t.Fatal(err)
+			}
+			_, err = client.AcquireTokenSilent(context.Background(), tokenScope)
+			// first attempt should fail
+			if err == nil {
+				t.Fatal("unexpected nil error from AcquireTokenSilent")
+			}
+			tk, err := client.AcquireTokenByAuthCode(context.Background(), "fake_auth_code", "fake_redirect_uri", tokenScope)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if tk.AccessToken != token {
+				t.Fatalf("unexpected access token %s", tk.AccessToken)
+			}
+			account := client.Account(tk.Account.HomeAccountID)
+			if params.utid == "" {
+				if account.HomeAccountID != "123-456.123-456" {
+					t.Fatal("Unexpected Account.HomeAccountId")
+				}
```

Comment:
Just a minor nitpick--including the expected and actual values in failure messages helps future debuggers figure out what went wrong:
```suggestion
				if actual := account.HomeAccountID; actual != "123-456.123-456" {
					t.Fatalf("expected %q, got %q", "123-456.123-456", actual)
				}
```
(tidier if the expected value comes from a variable instead of a literal)